### DISCUSSION TITLE
XAIO ESP32S3 & Wio_SX1262 connect by their onboard B2B connector

### DISCUSSION THREAD

**Discussion by ctmorrison** [jgromes/RadioLib#1361] [Q&A]
I can't seem to figure out how to get a simple transmit and receive Arduino application working with these two boards I recently purchased.  Can anyone offer some insight?  I imagine it's associated with the use of the B2B connector (I think they call it Sense), but I'm not sure.  If it would work better if I not use the connector and breadboard them, please advise.

**Comment by jgromes** [comment]
I don't know either of those, and I don't what you mean by B2B connector. The library doesn't care what kind of connector you use, just that you specify the correct SPI and digital pins.

**Reply by StevenCellist** [reply]
I have this board on the desk waiting for an attempt. I will give it a go after Christmas. Happy holidays!

**Reply by ctmorrison** [reply]
The two boards are connected by a B2B connector.  What I haven't been able to determine which pin is which on the B2B connector so I can make sure I'm specifying the correct SPI pins when I instantiate the radio object.  I'm getting conflicting information.

**Reply by ctmorrison** [reply]
I just want to implement simple LoRa communications between two devices, each of which uses these same pairs of boards connected via the B2B connectors

**Comment by radiohound** [comment]
The pinout for their board to board connector is on [this page](https://files.seeedstudio.com/wiki/SeeedStudio-XIAO-ESP32S3/res/XIAO_ESP32S3_ExpBoard_v1.0_SCH.pdf)

and [here](https://files.seeedstudio.com/products/SenseCAP/Wio_SX1262/Schematic_Diagram_Wio-SX1262_for_XIAO.pdf) for the b2b version of the WIO-SX1262

**Comment by ctmorrison** [comment]
Thanks, Walter.  I believe I now know what pins to use when I instantiate the radio object.  Until now, I had conflicting information about which pins were used by the XAIO ESP32S3.  A bit of hallucination by ChatGPT also threw me off course.

Again, thank you!!!

**Comment by ctmorrison** [comment]
Any chance you can point me to the schematic for the Wio_SX1262 daughter card that plugs in via the B2B connector?  I need to confirm where all the pins are for the SPI communications.

**Comment by radiohound** [comment]
That schematic is [here](https://files.seeedstudio.com/products/SenseCAP/Wio_SX1262/Schematic_Diagram_Wio-SX1262_for_XIAO.pdf) for the b2b version of the WIO-SX1262

**Comment by ctmorrison** [comment]
Walter,

Again, thank you for your help!  However, I’m a bit confused about which connector is which.  The ESP32S3 does not show a 30 pin connector and the Wio_SX1262 does not show a 34 pin connector.  Can you provide any insight about what I’m missing?

Thanks,
Tom

**Reply by StevenCellist** [reply]
It's still Christmas so I'm surprised by the speed of this thread, but here's Seeed's pinmap for the OCH-firmware which seems to include a double pinmap:

https://github.com/Seeed-Studio/one_channel_hub/blob/4cc771ac02da1bd18be67509f6b52d21bb0feabd/components/smtc_ral/bsp/sx126x/seeed_xiao_esp32s3_devkit_sx1262.c#L358-L369

I assume it works, but haven't tried myself yet, so YMMV.

**Reply by radiohound** [reply]
I'm guessing the "34" pin connector will match the other pins, if you ignore 31, 32, 33 and 34. Their numbering conventions on both connectors are different as well. They should really fix those errors to reduce customers hair pulling.

![image](https://github.com/user-attachments/assets/74603b44-fea9-41e3-ac1d-003e7a9e356d)

![image](https://github.com/user-attachments/assets/3c800393-244b-4485-b632-f82914542000)

**Comment by ctmorrison** [comment]
I’ll try that to match things up.  Once again, thank you for your help.  I could not find those schematics.

Tom Morrison
2186 Talmage Drive
Leland, NC 28451
724-991-2473 (cell)
***@***.*** ***@***.***>

**Reply by mcsnoogins** [reply]
Did you get anywhere with this? Able to share the correct pins? I'm in exactly the same situation, matched up what i thought was right from the schematics and information from Steven but still not working :(

**Reply by StevenCellist** [reply]
Do you use the special B2B connector, or just the pins? For the B2B connector, support is available in RadioBoards, assuming you selected the XIAO ESP32-S3 as your board.

**Reply by mcsnoogins** [reply]
Hi, I'm using a xiao esp32s3 and wio-sx1262 kit connected using the b2b board. Will check out Radioboards thanks, have been going in circles with different pins for days :D

**Comment by ctmorrison** [comment]
Thanks, Steven.  I’ll need some time to absorb all of this, but at least I have some data with which to work!

Merry Christmas

**Comment by ctmorrison** [comment]
So I'm guessing at the following definitions I need to make in my XAIO ESP32S3 Arduino program:
MISO - D9
MOSI - D10
CS - D2
SCK - D8

But I still don't know how to assign pins to:
LORA RST
LORA NSS
LORA DIO1
LORA SW1
The documentation is seriously lacking and I'd be totally lost without the help on this form.  The SeeedStudio support here in the USA has been rather worthless, unfortunately.  How can it be so challenging if they want users to actually use these devices as they promote?  I know many use cases are easy.  However, simple LoRa communications seem to be a mystery.

**Reply by StevenCellist** [reply]
If you look at the link I shared, you see pin numbers for all pins you mentioned, with IRQ = DIO1 = GPIO1. So through the connector, NSS would be pin 41 for example, while through the pins, NSS would be pin 4. With that information, you should be able to get going I believe.

**Reply by StevenCellist** [reply]
(*) Correction: DIO1 only is connected to GPIO1 through the pins, for the connector its value is listed in that link.

**Reply by StevenCellist** [reply]
Support for the XIAO ESP32S3 has been added in https://github.com/radiolib-org/RadioBoards/commit/ee471e529720fb28d88dd4575064a4cb346dacb6.
It assumes the radio is connected using the B2B connector, which to me seems a fair assumption. This is also mentioned in the [warning message](https://github.com/radiolib-org/RadioBoards/blob/ee471e529720fb28d88dd4575064a4cb346dacb6/src/maintained/SeeedStudio/XIAO_ESP32S3.h#L10) when compiling. You can either download RadioBoards from GitHub and add it as a library, or clone it using `git clone`, and using the methods described at the start of all RadioLib examples. Or if you want to do neither of those, locate the `XIAO_ESP32S3.h` file and refer to the mentioned pin numbers there.

This really was just a matter of copying the pin numbers from the link I shared before. For next time, I advise you check out the resources that are shared with you (such as that link) before making stuff complicated by looking at (admittedly very poorly documented) Seeed schematics. That will hopefully save you some headaches ;)

**Comment by rees3901** [comment]
these pins worked for me (and the author)


https://forum.seeedstudio.com/t/xiao-esp32s3-wio-sx1262-lora-kit-as-p2p-lora-node/291513/2

**Reply by rees3901** [reply]
namely (excuse the tick emojis - not me !) :


include <SPI.h>
#include <RadioLib.h>
#include <OneWire.h>
#include <DallasTemperature.h>

// ✅ Updated Pin Configuration (B2B Connector)
#define LORA_NSS 41    // ✅ SPI Chip Select (GPIO41)
#define LORA_SCK 7     // ✅ SPI Clock (GPIO7)
#define LORA_MOSI 9    // ✅ SPI MOSI (GPIO9)
#define LORA_MISO 8    // ✅ SPI MISO (GPIO8)
#define LORA_RST 42    // ✅ LoRa Reset (GPIO42)
#define LORA_BUSY 40   // ✅ LoRa BUSY (GPIO40)
#define LORA_DIO1 39   // ✅ LoRa IRQ (DIO1 - GPIO39)
#define LORA_ANT_SW 38 // ✅ Antenna Switch (GPIO38)

**Reply by Reetabass** [reply]
Thank you so much! I was driving myself crazy trying to figure out why SPI was not working. Got it working with those pin configurations.

**Reply by rees3901** [reply]
glad I could be of use for once! 🤣

**Reply by UssamaZafar** [reply]
I am using theSx1262 module from wavesshare and i have made dev board embedded with ESP32S3, I am not very proficient with programming stuff but these are the pinouts I am using... Can anyone help me in configuring the radiolib.
#define LORA_NSS 21 // SPI Chip Select (GPIO21)
#define LORA_SCK 9 // SPI Clock (GPIO9)
#define LORA_MOSI 11 // SPI MOSI (GPI11)
#define LORA_MISO 10 //  SPI MISO (GPI10)
#define LORA_RST 14 // LoRa Reset (GPIO14)
#define LORA_BUSY 13 //  LoRa BUSY (GPIO13)
#define LORA_DIO1 46 // LoRa IRQ (DIO1 - GPIO46)

**Reply by Reetabass** [reply]
@UssamaZafar Sorry, I don't understand the question. Did you make your own PCB with the Waveshare SX1262 module and an ESP32-S3 MCU? Please advise and include schematics and datasheets for the modules you are using.

You would need to review the ESP32-S3 pinout and connect it to the LoRa module via SPI.

**Comment by PjGLasso** [comment]
Hi there,

So I was able to Use all of the test Software on The Wio-SX-1262 & Xiao ESP32S3 with B2B connector. (Board to Board)
Here is a **[link](https://forum.seeedstudio.com/t/using-the-wio-sx1262-for-xiao-to-easily-build-lora-point-to-point-communication-with-various-xiaos/291605/21?u=pj_glasso)** to the thread on the support form with working code.
I edited your Ping/Pong with 
`// SX1262 radio = new Module(10, 2, 3, 9);
SX1262 radio = new Module(41, 39, 42, 40); //support for the Wio_SX-1262 & Xiao ESP32S3 with B2B connector.
`

Everything is working Perfectly. :+1:
HTH
GL :-) PJ :v:

**Reply by goblingift** [reply]
You reply is golden! It should be mentioned on the official Seeed wiki.

**Comment by DareDareHub** [comment]
For those using the Xiao ESP32-S3 and the WIO SX1262 with the pins (not with the B2B connector) here is what worked for me using the RadioLib example from https://github.com/jgromes/RadioLib
Examples > RadioLib > SX126x > SX126x_PingPong.ino

Set the pins to match the WIO-SX62 module (use this schematic to match and not the drawing from the product page on  SEEDSTUDIO)
https://files.seeedstudio.com/products/SenseCAP/Wio_SX1262/Wio-SX1262%20for%20XIAO%20V1.0_SCH.pdf
```
// SX1262 has the following connections:
// NSS pin: 5
// DIO1 pin: 2
// NRST pin: 3
// BUSY pin: 4
SX1262 radio = new Module(5, 2, 3, 4);
```
Serial Monitor result :
15:48:46.633 -> [SX1262] Starting to listen ... success!

**Comment by Reetabass** [comment]
I got the SX1262x working with ESP32S3 in ESPIDF.

If you are using Arduino, you can include radiolib.h and Arduino will handle the imports for you. I am using ESP-IDF to learn more about it, and have been using Arduino Core to import code written in Arduino.

I was initially debugging SX1262 init failures (SPI command errors, TX timeouts) and assumed it was a RadioLib or ESP-IDF integration issue. The real problem turned out to be incorrect pin assumptions.

The pins for the Seed studio b2b connector are: 
#define LORA_NSS 41 // ✅ SPI Chip Select (GPIO41)
#define LORA_SCK 7 // ✅ SPI Clock (GPIO7)
#define LORA_MOSI 9 // ✅ SPI MOSI (GPIO9)
#define LORA_MISO 8 // ✅ SPI MISO (GPIO8)
#define LORA_RST 42 // ✅ LoRa Reset (GPIO42)
#define LORA_BUSY 40 // ✅ LoRa BUSY (GPIO40)
#define LORA_DIO1 39 // ✅ LoRa IRQ (DIO1 - GPIO39)
#define LORA_ANT_SW 38 // ✅ Antenna Switch (GPIO38) 

Thanks to @rees3901 for those correct pins

Once I aligned my menuconfig pin assignments with the B2B connector mapping, SX1262 initialisation succeeded, and LoRa TX worked reliably.

Also, the SX1262x from Seed Studio uses a TCXO at 1.8V, so you have to set that as well.

[https://files.seeedstudio.com/products/SenseCAP/Wio_SX1262/Wio-SX1262_Module_Datasheet.pdf](url)

Here is the repo if you are interested in looking at it.

[https://github.com/Reetabass/RawLora_SeedStudio_ESP32S3](url)
